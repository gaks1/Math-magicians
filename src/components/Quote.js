import React, { useState, useEffect } from 'react';

// fetch data from https://api.api-ninjas.com/v1/quotes?category=success api using fetch
function Quote() {
  const [quote, setQuote] = useState('');

  useEffect(() => {

    const fetchData = async () => {
      try {
        const response = await fetch(
          'https://api.api-ninjas.com/v1/quotes?category=success',
          {
            method: 'GET',
            headers: {
              'X-Api-Key': 'diF2CPTnfXZ3z+eeylqvtg==6y2p5KUk1d9qZPsk',
            },
          },
        );

        const data = await response.json();
        setQuote(data[0].quote);
      } catch (error) {
      }
    };

    fetchData();

  }, []);

  return (
    <div className="quote">
      <h1>Quote of the day</h1>
      <p>{quote}</p>
    </div>
  );
}

export default Quote;
