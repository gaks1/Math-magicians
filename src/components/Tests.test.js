import { render, fireEvent, screen } from '@testing-library/react';
import React from 'react';
import renderer from 'react-test-renderer';
import calculate from '../logic/Calculate';
import operate from '../logic/Operate';
import '@testing-library/jest-dom';
import Calculator from './Calculator';
import Quote from './Quote';

describe('Calculator', () => {
  it('renders correctly', () => {
    const tree = renderer.create(<Calculator />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe('Calculator component', () => {
  beforeEach(() => {
    render(<Calculator />);
  });

  test('renders calculator with initial state', () => {
    expect(screen.getByText("Let's do some math!")).toBeInTheDocument();
    expect(screen.getByTestId('display').textContent).toBe('0');
  });

  test('renders calculator with AC', () => {
    fireEvent.click(
      screen.getAllByText('1').find((el) => el.tagName === 'BUTTON')
    );
    fireEvent.click(
      screen.getAllByText('AC').find((el) => el.tagName === 'BUTTON')
    );
    expect(screen.getByTestId('display').textContent).toBe('0');
  });

  test('renders calculator with 1', () => {
    fireEvent.click(
      screen.getAllByText('1').find((el) => el.tagName === 'BUTTON')
    );
    expect(screen.getByTestId('display').textContent).toBe('1');
  });

  test('renders calculator with 1 + 1', () => {
    fireEvent.click(
      screen.getAllByText('1').find((el) => el.tagName === 'BUTTON')
    );
    fireEvent.click(
      screen.getAllByText('+').find((el) => el.tagName === 'BUTTON')
    );
    fireEvent.click(
      screen.getAllByText('1').find((el) => el.tagName === 'BUTTON')
    );
    fireEvent.click(
      screen.getAllByText('=').find((el) => el.tagName === 'BUTTON')
    );
    expect(screen.getByTestId('display').textContent).toBe('2');
  });
});

describe('Calculate function', () => {
  test('AC button resets the calculator', () => {
    const obj = {
      total: '5',
      next: '3',
      operation: '+',
    };
    const result = calculate(obj, 'AC');
    expect(result).toEqual({ total: null, next: null, operation: null });
  });

  test('Decimal button adds decimal point', () => {
    const obj = {
      total: null,
      next: '5',
      operation: null,
    };
    const result = calculate(obj, '.');
    expect(result).toEqual({ total: null, next: '5.', operation: null });
  });

  test('= button performs the operation', () => {
    const obj = {
      total: '5',
      next: '3',
      operation: '+',
    };
    const result = calculate(obj, '=');
    expect(result).toEqual({ total: '8', next: null, operation: null });
  });

  test('+/- button negates the value', () => {
    const obj = {
      total: '5',
      next: '3',
      operation: '+',
    };
    const result = calculate(obj, '+/-');
    expect(result).toEqual({ total: '5', next: '-3', operation: '+' });
  });

  test('Operation button sets the operation', () => {
    const obj = {
      total: '5',
      next: null,
      operation: null,
    };
    const result = calculate(obj, '+');
    expect(result).toEqual({ total: '5', next: null, operation: '+' });
  });
});

describe('operate', () => {
  test('adding', () => {
    expect(operate(1, 2, '+')).toBe('3');
  });

  test('subtracting', () => {
    expect(operate(1, 2, '-')).toBe('-1');
  });

  test('multiplying', () => {
    expect(operate(1, 2, 'x')).toBe('2');
  });

  test('dividing', () => {
    expect(operate(1, 2, 'รท')).toBe('0.5');
  });

  test('dividing by zero', () => {
    expect(operate(1, 0, 'รท')).toBe("Can't divide by 0.");
  });

  test('modulus', () => {
    expect(operate(1, 2, '%')).toBe('1');
  });

  test('modulus by zero', () => {
    expect(operate(1, 0, '%')).toBe("Can't find modulo as can't divide by 0.");
  });

  test('invalid operator', () => {
    let errorThrown = false;
    try {
      operate(1, 2, '/');
    } catch (err) {
      errorThrown = true;
      expect(err.message).toBe("Unknown operation '/'");
    }
    expect(errorThrown).toBe(true);
  });

  test('matches snapshot', () => {
    const { asFragment } = render(<Quote />);
    expect(asFragment()).toMatchSnapshot();
  });
  test('renders quote after loading data', async () => {
    const mockData = [{ quote: 'Success is not final, failure is not fatal' }];
    jest.spyOn(window, 'fetch').mockImplementationOnce(() => {
      return Promise.resolve({
        json: () => Promise.resolve(mockData),
      });
    });
    render(<Quote />);
    const quote = await screen.findByText(mockData[0].quote);
    expect(quote).toBeInTheDocument();
    window.fetch.mockRestore();
  });
});
