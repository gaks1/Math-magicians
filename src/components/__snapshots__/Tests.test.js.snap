// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calculator renders correctly 1`] = `
<div
  className="calcPage"
>
  <h2>
    Let's do some math!
  </h2>
  <div
    className="calculator"
  >
    <div
      data-testid="display"
      id="screen"
    >
      0
    </div>
    <div
      className="row"
    >
      <button
        className="left-button"
        onClick={[Function]}
        type="button"
      >
        AC
      </button>
      <button
        className="left-button"
        onClick={[Function]}
        type="button"
      >
        +/-
      </button>
      <button
        className="left-button"
        onClick={[Function]}
        type="button"
      >
        %
      </button>
      <button
        className="right-button"
        onClick={[Function]}
        type="button"
      >
        รท
      </button>
    </div>
    <div
      className="row"
    >
      <button
        className="left-button"
        onClick={[Function]}
        type="button"
      >
        7
      </button>
      <button
        className="left-button"
        onClick={[Function]}
        type="button"
      >
        8
      </button>
      <button
        className="left-button"
        onClick={[Function]}
        type="button"
      >
        9
      </button>
      <button
        className="right-button"
        onClick={[Function]}
        type="button"
      >
        x
      </button>
    </div>
    <div
      className="row"
    >
      <button
        className="left-button"
        onClick={[Function]}
        type="button"
      >
        4
      </button>
      <button
        className="left-button"
        onClick={[Function]}
        type="button"
      >
        5
      </button>
      <button
        className="left-button"
        onClick={[Function]}
        type="button"
      >
        6
      </button>
      <button
        className="right-button"
        onClick={[Function]}
        type="button"
      >
        -
      </button>
    </div>
    <div
      className="row"
    >
      <button
        className="left-button"
        onClick={[Function]}
        type="button"
      >
        1
      </button>
      <button
        className="left-button"
        onClick={[Function]}
        type="button"
      >
        2
      </button>
      <button
        className="left-button"
        onClick={[Function]}
        type="button"
      >
        3
      </button>
      <button
        className="right-button"
        onClick={[Function]}
        type="button"
      >
        +
      </button>
    </div>
    <div
      className="row"
    >
      <button
        className="left-button zero"
        onClick={[Function]}
        type="button"
      >
        0
      </button>
      <button
        className="left-button dot"
        onClick={[Function]}
        type="button"
      >
        .
      </button>
      <button
        className="right-button"
        onClick={[Function]}
        type="button"
      >
        =
      </button>
    </div>
  </div>
</div>
`;

exports[`operate matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="quote"
  >
    <p
      class="quote"
    />
    <p>
      Loading...
    </p>
  </div>
</DocumentFragment>
`;
